mitodl/mitxpro
# Support reentry to auth flow for users who abandon registration

#### Pre-Flight checklist


#### What are the relevant tickets?
Fixes #841 
Fixes #867
Closes #312 

#### What's this PR do?

Generally speaking we're trying to round off edge cases where the user can get into weird states with regards to registration flow abandonment.

Specifically it:

- Updates `Profile` fields to be nullable/blankable
- Backpopulates `Profile`
- Moves creation of `Profile` to within the transaction of creating `User`
- Moves when we enable the user to right after we verify them with exports
- Updates the frontend to support entry to the extra details page from the login flow
- Fills in the remaining tests around login/registration pages

There were a few knock-on fixes required in unrelated modules that didn't have solid tests.

#### How should this be manually tested?

This requires a lot of permutation testing. A good start is:

- Register a new user normally
- Login with an existing user normally
- Abandon the registration flow after navigating to the extra details page and attempt to login again (you should be redirected back to the extra details step)
- Register with a user who fails exports control (should get the error page)
- Attempt to login with a user who fails exports control (should get the error page and *not* be logged in or prompted to enter any information)
- Attempt to register with an existing user
- Attempt to login with a nonexistent user
